nemo/collections/nlp/modules/common/hyena/hyena.py

```diff
@@ -0,0 +1,386 @@
+
+
+class HyenaConv(nn.Module):
+    def __init__(
+        self,
+        d_model: int,
+        l_max: int,
+        order: int,
+        filter_order: int,
+        bias: bool = True,
+        filter_cls: Union[ModuleSpec, type] = HyenaFilter,
+        filter_submodules: HyenaFilterSubmodules = None,
+        **filter_kwargs,
+    ):
+        super().__init__()
+        self.d_model = d_model
+        self.order = order
+        self.l_max = l_max
+        self.use_bias = bias
+        self.bias = nn.Parameter(torch.randn(self.d_model * (self.order - 1))) if self.use_bias else None
+
+        self.filter = build_module(
+            filter_cls,
+            self.d_model * (self.order - 1),
+            submodules=filter_submodules,
+            order=filter_order,
+            seq_len=l_max,
+            **filter_kwargs,
+        )
+
+
+class SingleHeadHyenaConv(HyenaConv):
+    def __init__(
+        self,
+        d_model: int,
+        l_max: int,
+        order: int,
+        filter_order: int,
+        bias: bool = True,
+        filter_cls: Union[ModuleSpec, type] = HyenaFilter,
+        filter_submodules: HyenaFilterSubmodules = None,
+        fftconv_type: str = None,
+        precision: str = 'bf16',
+        **filter_kwargs,
+    ):
+        super().__init__(
+            d_model,
+            l_max,
+            order,
+            filter_order,
+            bias=bias,
+            filter_cls=filter_cls,
+            filter_submodules=filter_submodules,
+            **filter_kwargs,
+        )
+
+        if fftconv_type is None:
+            if l_max <= 8192 and HAVE_FFTCONV:
+                # safari-fftconv supports seq-len <= 8192 and is a bit faster vs. flashfftconv
+                fftconv_type = 'safari'
+            else:
+                fftconv_type = 'flash'
+
+        if fftconv_type not in ['safari', 'flash']:
+            raise ValueError("fftconv_type must be one of ['safari', 'flash']")
+        if fftconv_type == 'safari' and not HAVE_FFTCONV:
+            raise ImportError('fftconv library not found. Please see README at <tbd> for instructions.')
+        if fftconv_type == 'flash' and not HAVE_FLASHFFTCONV:
+            raise ImportError('flashfftconv library not found. Please see README at <tbd> for instructions.')
+
+        if fftconv_type == 'safari':
+            self.fftconv_fn = self._safari_fft
+        else:  # fftconv_type == 'flash'
+            self.flashfftconv = FlashFFTConv(2 * self.l_max, torch_dtype_from_precision(precision)).flashfftconv
+            self.fftconv_fn = self._flash_fft
+
+    def _safari_fft(self, x, k, bias):
+        bias = bias.to(dtype=torch.float32)
+        return safari_fftconv_fn(x, k, bias, gelu=False)
+
+    def _flash_fft(self, x, k, bias):
+        x = x.contiguous()
+        y = self.flashfftconv(x, k) + x * bias.unsqueeze(dim=1)
+        return y
+
+    def forward(self, x, k, recurrence_idx):
+        bias = self.bias if self.use_bias else 0 * self.bias
```

Comment:
is there any way to avoid the 0 * self.bias? I wonder if changing fftconv_fn to accept bias=None would be possible here?

Reply:
safari-fftconv expects a valid tensor. But definitely no reason to do this on every forward pass, moved it to `__init__`, creating a "vanilla" zero tensor (that is - not a `nn.Parameter`).